pkg/azurefile/controllerserver.go

```diff
@@ -987,6 +990,114 @@ func (d *Driver) ListSnapshots(_ context.Context, _ *csi.ListSnapshotsRequest) (
+	azcopyCopyOptions := defaultAzcopyCopyOptions
+	if snapshot != "" {
+		srcFileShareName = fmt.Sprintf("%s(snapshot: %s)", srcFileShareName, snapshot)
+		azcopyCopyOptions = snapshotAzcopyCopyOptions
+	}
+
+	jobState, percent, err := d.azcopy.GetAzcopyJob(dstFileShareName, authAzcopyEnv)
+	klog.V(2).Infof("azcopy job status: %s, copy percent: %s%%, error: %v", jobState, percent, err)
+	switch jobState {
+	case fileutil.AzcopyJobError, fileutil.AzcopyJobCompleted:
+		return err
+	case fileutil.AzcopyJobRunning:
+		return fmt.Errorf("an existed azcopy job is running, copy percent: %s%%, please wait for it to complete", percent)
+	case fileutil.AzcopyJobNotFound:
+		klog.V(2).Infof("copy fileshare %s to %s", srcFileShareName, dstFileShareName)
+		out, copyErr := d.waitForAzcopyCompletion(srcPath, dstPath, srcFileShareName, dstFileShareName, azcopyCopyOptions, authAzcopyEnv, time.Duration(d.waitForAzCopyTimeoutMinutes)*time.Minute, d.execAzcopyCopy)
+		if accountSASToken == "" && strings.Contains(string(out), authorizationPermissionMismatch) && copyErr != nil {
+			klog.Warningf("azcopy list failed with AuthorizationPermissionMismatch error, should assign \"Storage File Data SMB Share Elevated Contributor\" role to controller identity, fall back to use sas token, original output: %v", string(out))
+			d.azcopySasTokenCache.Set(accountName, "")
+			var sasToken string
+			if sasToken, _, err = d.getAzcopyAuth(ctx, accountName, "", storageEndpointSuffix, accountOptions, secrets, secretName, secretNamespace, true); err != nil {
+				return err
+			}
+			out, copyErr = d.waitForAzcopyCompletion(srcPath+sasToken, dstPath+sasToken, srcFileShareName, dstFileShareName, azcopyCopyOptions, []string{}, time.Duration(d.waitForAzCopyTimeoutMinutes)*time.Minute, d.execAzcopyCopy)
+		}
+		if copyErr != nil {
+			klog.Warningf("CopyFileShare(%s, %s, %s) failed with error(%v): %v", resourceGroupName, accountName, dstFileShareName, copyErr, string(out))
+		} else {
+			klog.V(2).Infof("copied fileshare %s to %s successfully", srcFileShareName, dstFileShareName)
+		}
+		return copyErr
+	}
+	return err
+}
+
+// execAzcopyCopy exec azcopy copy command
+func (d *Driver) execAzcopyCopy(srcPath, dstPath string, azcopyCopyOptions, authAzcopyEnv []string) ([]byte, error) {
+	cmd := exec.Command("azcopy", "copy", srcPath, dstPath)
+	cmd.Args = append(cmd.Args, azcopyCopyOptions...)
+	if len(authAzcopyEnv) > 0 {
+		cmd.Env = append(os.Environ(), authAzcopyEnv...)
+	}
+	return cmd.CombinedOutput()
+}
+
+// waitForAzcopyCompletion waits for the exec azcopy copy function to complete or times out
+func (d *Driver) waitForAzcopyCompletion(srcPath, dstPath, srcFileShareName, dstFileShareName string, azcopyCopyOptions, authAzcopyEnv []string, timeout time.Duration, f func(string, string, []string, []string) ([]byte, error)) ([]byte, error) {
```

Comment:
`f` is too simple, use `azCopyFunc` instead

Reply:
done